File: verify_port.mjs
/**
 * Port parity check — Node engine vs Python backtest.
 *
 * §6.2 commits the production compute job to reusing Meridian's JS functions
 * rather than reimplementing them in Python, precisely to avoid the
 * dual-codebase drift §7.2 caught in a third-party scorer. This proves the
 * commitment actually holds: it runs meridian-engine.js under Node over the
 * real price history and emits the Golden Breakout candidate set for the
 * latest date, which verify_port.py then compares against the independent
 * Python implementation in meridian_backtest.py over the same data.
 *
 * Two implementations, two languages, one expected answer. Any disagreement is
 * drift, and is meant to fail loudly here rather than quietly in production.
 *
 * Usage: node verify_port.mjs > port-node.json
 */

import { readFileSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { computeAll, runGoldenBreakoutScreener } from "./meridian-engine.js";

const BASE = dirname(fileURLToPath(import.meta.url));

// Minimal RFC4180-ish parser: 75 rows of the master carry quoted fields with
// embedded commas ("Food, Beverages & Tobacco"), so splitting on "," is wrong.
function parseCSV(text) {
  const rows = [];
  let field = "", row = [], inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') { field += '"'; i++; } else inQuotes = false;
      } else field += c;
    } else if (c === '"') inQuotes = true;
    else if (c === ",") { row.push(field); field = ""; }
    else if (c === "\n") { row.push(field); rows.push(row); row = []; field = ""; }
    else if (c !== "\r") field += c;
  }
  if (field || row.length) { row.push(field); rows.push(row); }
  const header = rows.shift();
  return rows
    .filter((r) => r.length === header.length && r.some((v) => v !== ""))
    .map((r) => Object.fromEntries(header.map((h, i) => [h, r[i]])));
}

const num = (v) => (v === "" || v == null ? null : Number(v));

const master = parseCSV(readFileSync(join(BASE, "meridian-company-master-2138.csv"), "utf8"))
  .map((m) => ({ ...m, MarketCap: num(m.MarketCap) }));

const priceFiles = readdirSync(BASE)
  .filter((f) => /^meridian-price-history-2090-part\d+of3\.csv$/.test(f))
  .sort();

const prices = [];
for (const f of priceFiles) {
  for (const r of parseCSV(readFileSync(join(BASE, f), "utf8"))) {
    prices.push({
      ISIN: r.ISIN, Date: r.Date,
      High: num(r.High), Low: num(r.Low),
      Close: num(r.Close), Volume: num(r.Volume),
    });
  }
}

// Fundamentals are irrelevant to the Golden Breakout gates (§4.3 keeps the
// Composite Score display-only), so the comparison runs technicals-only —
// exactly the surface meridian_backtest.py implements.
const computed = computeAll(master, [], prices);
const candidates = runGoldenBreakoutScreener(computed);

const asOf = prices.reduce((m, r) => (r.Date > m ? r.Date : m), "");
process.stdout.write(JSON.stringify({
  as_of: asOf,
  instruments: new Set(prices.map((r) => r.ISIN)).size,
  candidates: candidates.map((c) => ({
    isin: c.ISIN,
    separation_pct: c.tech.separationPct,
    freshness_days: c.tech.goldenCrossStreak?.streak ?? null,
    ma200_slope_pct: c.tech.ma200SlopePct,
  })),
}, null, 1) + "\n");

File: verify_port.py
"""
Port parity check — compares the Node engine's Golden Breakout output against
the independent Python implementation in meridian_backtest.py.

Run:
    node verify_port.mjs > port-node.json
    python3 verify_port.py

Exits non-zero on any disagreement, so this can gate a build.
"""

import json
import os
import sys

import pandas as pd

from meridian_backtest import (
    compute_all_signals,
    load_price_data_from_csv,
    passes_golden_breakout,
)

BASE = os.path.dirname(os.path.abspath(__file__))
NODE_OUT = os.path.join(BASE, "port-node.json")

if not os.path.exists(NODE_OUT):
    sys.exit("run `node verify_port.mjs > port-node.json` first")

node = json.load(open(NODE_OUT))
node_set = {c["isin"] for c in node["candidates"]}
node_by = {c["isin"]: c for c in node["candidates"]}

close = load_price_data_from_csv(os.path.join(BASE, "meridian-price-history-2090-part*of3.csv"))
signals = compute_all_signals(close)
sig = passes_golden_breakout(signals)

as_of = sig.index.max()
py_set = set(sig.loc[as_of][sig.loc[as_of]].index)

print(f"as-of date      node={node['as_of']}  python={as_of.date()}")
print(f"instruments     node={node['instruments']}  python={close.shape[1]}")
print(f"candidates      node={len(node_set)}  python={len(py_set)}")

only_node = node_set - py_set
only_py = py_set - node_set
shared = node_set & py_set
print(f"agreement       {len(shared)} shared, {len(only_node)} node-only, {len(only_py)} python-only")

if shared:
    print("\nper-candidate numeric agreement (separation %, freshness days):")
    sep = signals["separation_pct"].loc[as_of]
    fresh = signals["golden_streak"].loc[as_of]
    worst = 0.0
    for isin in sorted(shared):
        n = node_by[isin]
        d_sep = abs(n["separation_pct"] - float(sep[isin]))
        d_fresh = abs(n["freshness_days"] - float(fresh[isin]))
        worst = max(worst, d_sep)
        flag = "" if d_sep < 1e-6 and d_fresh == 0 else "   <-- MISMATCH"
        print(f"  {isin}  sep node={n['separation_pct']:.6f} py={float(sep[isin]):.6f} "
              f"(d={d_sep:.2e})  fresh node={n['freshness_days']} py={int(fresh[isin])}{flag}")
    print(f"\nlargest separation difference: {worst:.2e}")

if only_node or only_py:
    print("\nDISAGREEMENT")
    for i in sorted(only_node):
        print(f"  node-only:   {i}")
    for i in sorted(only_py):
        print(f"  python-only: {i}")
    sys.exit(1)

print("\nPARITY OK — both implementations select the identical candidate set.")
